src/listener.rs

```diff
@@ -0,0 +1,237 @@
+
+use crate::StatusCode;
+use async_lock::Semaphore;
+use async_std::{
+    net::TcpListener,
+    sync::Arc,
+    task::{sleep, spawn},
+};
+use async_trait::async_trait;
+use derivative::Derivative;
+use futures::stream::StreamExt;
+use std::{
+    fmt::{self, Display, Formatter},
+    io::{self, ErrorKind},
+    net::SocketAddr,
+    time::Duration,
+};
+use tide::{
+    http,
+    listener::{ListenInfo, Listener, ToListener},
+    Server,
+};
+
+/// TCP listener which accepts only a limited number of connections at a time.
+///
+/// This listener is based on [`tide::listener::TcpListener`] and should match the semantics of that
+/// listener in every way, accept that when there are more simultaneous outstanding requests than
+/// the configured limit, excess requests will fail immediately with error code 429 (Too Many
+/// Requests).
+#[derive(Derivative)]
+#[derivative(Debug(bound = "State: Send + Sync + 'static"))]
+pub struct RateLimitListener<State> {
+    addr: SocketAddr,
+    listener: Option<TcpListener>,
+    server: Option<Server<State>>,
+    info: Option<ListenInfo>,
+    permit: Arc<Semaphore>,
+}
+
+impl<State> RateLimitListener<State> {
+    /// Listen at the given address.
+    pub fn new(addr: SocketAddr, limit: usize) -> Self {
+        Self {
+            addr,
+            listener: None,
+            server: None,
+            info: None,
+            permit: Arc::new(Semaphore::new(limit)),
+        }
+    }
+
+    /// Listen at the given port on all interfaces.
+    pub fn with_port(port: u16, limit: usize) -> Self {
+        Self::new(([0, 0, 0, 0], port).into(), limit)
+    }
+}
+
+#[async_trait]
+impl<State> Listener<State> for RateLimitListener<State>
+where
+    State: Clone + Send + Sync + 'static,
+{
+    async fn bind(&mut self, app: Server<State>) -> io::Result<()> {
+        if self.server.is_some() {
+            return Err(io::Error::new(
+                ErrorKind::AlreadyExists,
+                "`bind` should only be called once",
+            ));
+        }
+        self.server = Some(app);
+        self.listener = Some(TcpListener::bind(&[self.addr][..]).await?);
+
+        // Format the listen information.
+        let conn_string = format!("{}", self);
+        let transport = "tcp".to_owned();
+        let tls = false;
+        self.info = Some(ListenInfo::new(conn_string, transport, tls));
+
+        Ok(())
+    }
+
+    async fn accept(&mut self) -> io::Result<()> {
+        let server = self.server.take().ok_or_else(|| {
+            io::Error::other("`Listener::bind` must be called before `Listener::accept`")
+        })?;
+        let listener = self.listener.take().ok_or_else(|| {
+            io::Error::other("`Listener::bind` must be called before `Listener::accept`")
+        })?;
+
+        let mut incoming = listener.incoming();
+        while let Some(stream) = incoming.next().await {
+            match stream {
+                Err(err) if is_transient_error(&err) => continue,
```

Comment:
Are the only changes versus the normal listener the addition of the rate limiter?

Reply:
That is my intention. I literally copied the code from tide and tweaked it. Unfortunately I can't see any good way to do this as a wrapper because there's no interface I can wrap for "handle a single connection"